File: 09-flood-fill/09.py
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        startingColor = image[sr][sc]
        if startingColor == color:
            return image

        def coloring(image, sr, sc):
            if sr > -1 and sr < len(image) and sc > -1 and sc < len(image[0]) and image[sr][sc] == startingColor:
                image[sr][sc] = color
                coloring(image, sr + 1, sc)
                coloring(image, sr - 1, sc)
                coloring(image, sr, sc + 1)
                coloring(image, sr, sc - 1)
            return image
        return coloring(image, sr, sc)


# print(Solution().floodFill([[1, 1, 1], [1, 1, 0], [1, 0, 1]], 1, 1, 2))
# print(Solution().floodFill([[0, 0, 0], [0, 0, 0]], 0, 0, 0))
# print(Solution().floodFill([[0, 0, 0], [0, 0, 1]], 0, 0, 2))
# print(Solution().floodFill([[0, 0, 0], [0, 0, 0]], 1, 0, 2))
# print(Solution().floodFill([[0, 1, 0], [0, 0, 0]], 0, 0, 0))

File: 09-flood-fill/09.js
// https://leetcode.com/problems/flood-fill/
/**
 * @param {number[][]} image
 * @param {number} sr row
 * @param {number} sc column
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function (image, sr, sc, color, ignoreM = [[]]) {
    const initial = image[sr] && image[sr][sc] !== undefined ? image[sr][sc] : undefined;
    if (ignoreM[sr] === undefined) {
        ignoreM[sr] = [];
    }
    const ignore = ignoreM[sr][sc] === 1;
    ignoreM[sr][sc] = 1;

    if (!ignore) {
        if (image[sr] !== undefined && image[sr][sc - 1] === initial) {
            image = floodFill(image, sr, sc - 1, color, ignoreM);
        }

        if (image[sr] !== undefined && image[sr][sc + 1] === initial) {
            image = floodFill(image, sr, sc + 1, color, ignoreM);
        }

        if (image[sr - 1] !== undefined && image[sr - 1][sc] === initial) {
            image = floodFill(image, sr - 1, sc, color, ignoreM);
        }

        if (image[sr + 1] !== undefined && image[sr + 1][sc] === initial) {
            image = floodFill(image, sr + 1, sc, color, ignoreM);
        }
    }
    image[sr][sc] = color;
    return image;
};

console.log(
    floodFill(
        [
            [0, 1, 0],
            [0, 0, 0],
        ],
        0,
        0,
        0
    )
);

// console.log(
//     floodFill(
//         [
//             [0, 0, 0],
//             [0, 0, 1],
//         ],
//         0,
//         0,
//         2
//     )
// );
// console.log(
//     floodFill(
//         [
//             [1, 1, 1],
//             [1, 1, 0],
//             [1, 0, 1],
//         ],
//         1,
//         1,
//         2
//     )
// );

// console.log(
//     floodFill(
//         [
//             [0, 0, 0],
//             [0, 0, 0],
//         ],
//         0,
//         0,
//         0
//     )
// );
